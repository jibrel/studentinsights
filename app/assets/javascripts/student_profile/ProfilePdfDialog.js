import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import qs from 'query-string';
import Datepicker from '../components/Datepicker';
import {
  toSchoolYear,
  firstDayOfSchool
} from '../helpers/schoolYear';


// Render options for downloading a profile PDF
export default class ProfilePdfDialog extends React.Component {
  constructor(props, context) {
    super(props, context);

    const nowMoment = context.nowFn();
    this.state = {
      includeRestrictedNotes: false,
      filterFromDate: firstDayOfSchool(toSchoolYear(nowMoment)-1),
      filterToDate: nowMoment
    };

    this.checkboxContainerEl = null;
    this.onIncludeRestrictedNotesChanged = this.onIncludeRestrictedNotesChanged.bind(this);
    this.onFilterFromDateChanged = this.onFilterFromDateChanged.bind(this);
    this.onFilterToDateChanged = this.onFilterToDateChanged.bind(this);
    this.onClickGenerateStudentReport = this.onClickGenerateStudentReport.bind(this);
  }

  filterFromDateForQuery() {
    const {filterFromDate} = this.state;

    return filterFromDate.format('MM/DD/YYYY');
  }

  filterToDateForQuery() {
    const {filterToDate} = this.state;

    return filterToDate.format('MM/DD/YYYY');
  }

  studentReportURL() {
    const {studentId} = this.props;
    const {includeRestrictedNotes} = this.state;
    const filterFromDateForQuery = this.filterFromDateForQuery();
    const filterToDateForQuery = this.filterToDateForQuery();
    const checkboxEls = $(this.checkboxContainerEl).find('.ProfilePdfDialog-section').toArray();
    const sections = checkboxEls.filter(el => el.checked).map(el => el.value).join(',');
    const queryString = qs.stringify({
      sections: sections,
      from_date: filterFromDateForQuery,
      to_date: filterToDateForQuery,
      include_restricted_notes: includeRestrictedNotes
    });
    return `/students/${studentId}/student_report.pdf?${queryString}`;
  }

  onIncludeRestrictedNotesChanged(e) {
    this.setState({includeRestrictedNotes: e.target.checked});
  }

  onFilterFromDateChanged(dateText) {
    const textMoment = moment.utc(dateText, 'MM/DD/YYYY');
    const updatedMoment = (textMoment.isValid()) ? textMoment : null;
    this.setState({ filterFromDate: updatedMoment });
  }

  onFilterToDateChanged(dateText) {
    const textMoment = moment.utc(dateText, 'MM/DD/YYYY');
    const updatedMoment = (textMoment.isValid()) ? textMoment : null;
    this.setState({ filterToDate: updatedMoment });
  }

  onClickGenerateStudentReport(event) {
    window.location = this.studentReportURL();
    return null;
  }

  render() {
    const {allowRestrictedNotes, showTitle, style} = this.props;
    const {includeRestrictedNotes} = this.state;

    return (
      <div className="ProfilePdfDialog" style={{...styles.root, ...style}}>
        {showTitle && <h4 style={styles.title}>Student Report</h4>}
        <span style={styles.tableHeader}>Student privacy:</span>
        <div style={styles.optionsList}>
          <div style={styles.fixedWidthOption}>
            <label style={styles.optionLabel}>
              <input
                style={styles.optionCheckbox}
                type='checkbox'
                className="ProfilePdfDialog-include-restricted-notes"
                onChange={this.onIncludeRestrictedNotesChanged}
                value={includeRestrictedNotes}
              />
              Include restricted notes
            </label>
          </div>
        </div>
        <span style={styles.tableHeader}>Select sections to include in report:</span>
        <div style={styles.optionsList} ref={el => this.checkboxContainerEl = el}>
          {this.renderStudentReportSectionOption('notes','Notes')}
          {allowRestrictedNotes && (
            this.renderStudentReportSectionOption('restricted_notes','Restricted notes')
          )}
          {this.renderStudentReportSectionOption('services','Services')}
          {this.renderStudentReportSectionOption('attendance','Attendance')}
          {this.renderStudentReportSectionOption('discipline_incidents','Discipline Incidents')}
          {this.renderStudentReportSectionOption('assessments','Assessments')}
        </div>
        <span style={styles.tableHeader}>Select dates for the report:</span>
        <div style={styles.datesSection}>
          <div style={styles.dateBox}>
            <label style={styles.dateLabel}>From:</label>
            <Datepicker
              styles={{
                datepicker: styles.datepickerContainer,
                input: styles.datepickerInput
              }}
              value={this.state.filterFromDate.format('MM/DD/YYYY')}
              onChange={this.onFilterFromDateChanged}
              datepickerOptions={{
                showOn: 'both',
                dateFormat: 'mm/dd/yy',
                minDate: undefined
              }} />
          </div>
          <div style={styles.dateBox}>
            <label style={styles.dateLabel}>Until:</label>
            <Datepicker
              styles={{
                datepicker: styles.datepickerContainer,
                input: styles.datepickerInput
              }}
              value={this.state.filterToDate.format('MM/DD/YYYY')}
              onChange={this.onFilterToDateChanged}
              datepickerOptions={{
                showOn: 'both',
                dateFormat: 'mm/dd/yy',
                minDate: undefined
              }} />
          </div>
        </div>
        <div style={{display: 'flex', justifyContent: 'flex-end'}}>
          <button
            style={styles.studentReportButton}
            className="btn btn-warning"
            onClick={this.onClickGenerateStudentReport}>
            Generate Student Report
          </button>
        </div>
      </div>
    );
  }

  renderStudentReportSectionOption(optionValue, optionName) {
    return (
      <div style={styles.fixedWidthOption}>
        <label style={styles.optionLabel}>
          <input
            style={styles.optionCheckbox}
            type='checkbox'
            className="ProfilePdfDialog-section"
            name={optionValue}
            defaultChecked
            value={optionValue} />
          {optionName}
        </label>
      </div>
    );
  }
}
ProfilePdfDialog.contextTypes = {
  nowFn: PropTypes.func.isRequired
};
ProfilePdfDialog.propTypes = {
  studentId: PropTypes.number.isRequired,
  allowRestrictedNotes: PropTypes.bool.isRequired,
  showTitle: PropTypes.bool,
  style: PropTypes.object
};

const styles = {
  root: {
    display: 'flex',
    flexDirection: 'column'
  },
  title: {
    borderBottom: '1px solid #333',
    color: 'black',
    padding: 10,
    paddingLeft: 0,
    marginBottom: 10
  },
  tableHeader: {
    fontWeight: 'bold',
    textAlign: 'left',
    marginBottom: 10
  },
  optionsList: {
    paddingLeft: 10,
    paddingBottom: 10
  },
  fixedWidthOption: {
    display: 'inline-block',
    width: '12em',
    height: '1.8em',
    verticalAlign: 'middle'
  },
  optionCheckbox: {
    marginRight: 5
  },
  optionLabel: {
    display: 'inline-block'
  },
  datesSection: {
    display: 'flex',
    marginLeft: 10,
    flexDirection: 'column'
  },
  dateBox: {
    display: 'flex',
    alignItems: 'center'
  },
  dateLabel: {
    display: 'block',
    width: '4em',
    marginBottom: 0
  },
  datepickerContainer: {
    display: 'block'
  },
  datepickerInput: {
    fontSize: 14,
    width: '8em',
    padding: 5
  },
  studentReportButton: {
    fontSize: 12,
    margin: 0
  }
};
