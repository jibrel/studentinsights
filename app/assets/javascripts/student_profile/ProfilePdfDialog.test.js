import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import {withDefaultNowContext, testContext} from '../testing/NowContainer';
import ProfilePdfDialog from './ProfilePdfDialog';

function testProps(props = {}) {
  return {
    studentId: 42,
    showTitle: true,
    allowRestrictedNotes: false,
    style: {},
    ...props
  };
}

function testEl(props = {}) {
  return withDefaultNowContext(<ProfilePdfDialog {...props} />);
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(testEl(props), el);
});

it('snapshots view', () => {
  const props = testProps();
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('PDF', () => {
  it('creates download URL in correct format', () => {
    const context = testContext();
    const wrapper = mount(<ProfilePdfDialog {...testProps()} />, {context});
    expect(wrapper.instance().studentReportURL()).toEqual('/students/42/student_report.pdf?from_date=08%2F15%2F2016&include_restricted_notes=false&sections=notes%2Cservices%2Cattendance%2Cdiscipline_incidents%2Cassessments&to_date=03%2F13%2F2018');
  });

  it('creates download URL in correct format with include_restricted_notes', () => {
    const context = testContext();
    const wrapper = mount(<ProfilePdfDialog {...testProps()} />, {context});
    wrapper.setState({includeRestrictedNotes: true});
    expect(wrapper.instance().studentReportURL()).toEqual('/students/42/student_report.pdf?from_date=08%2F15%2F2016&include_restricted_notes=true&sections=notes%2Cservices%2Cattendance%2Cdiscipline_incidents%2Cassessments&to_date=03%2F13%2F2018');
  });
});
