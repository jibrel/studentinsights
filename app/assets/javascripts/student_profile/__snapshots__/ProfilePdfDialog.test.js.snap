// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots view 1`] = `
<div
  className="ProfilePdfDialog"
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <h4
    style={
      Object {
        "borderBottom": "1px solid #333",
        "color": "black",
        "marginBottom": 10,
        "padding": 10,
        "paddingLeft": 0,
      }
    }
  >
    Student Report
  </h4>
  <span
    style={
      Object {
        "fontWeight": "bold",
        "marginBottom": 10,
        "textAlign": "left",
      }
    }
  >
    Student privacy:
  </span>
  <div
    style={
      Object {
        "paddingBottom": 10,
        "paddingLeft": 10,
      }
    }
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "height": "1.8em",
          "verticalAlign": "middle",
          "width": "12em",
        }
      }
    >
      <label
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          className="ProfilePdfDialog-include-restricted-notes"
          onChange={[Function]}
          style={
            Object {
              "marginRight": 5,
            }
          }
          type="checkbox"
          value={false}
        />
        Include restricted notes
      </label>
    </div>
  </div>
  <span
    style={
      Object {
        "fontWeight": "bold",
        "marginBottom": 10,
        "textAlign": "left",
      }
    }
  >
    Select sections to include in report:
  </span>
  <div
    style={
      Object {
        "paddingBottom": 10,
        "paddingLeft": 10,
      }
    }
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "height": "1.8em",
          "verticalAlign": "middle",
          "width": "12em",
        }
      }
    >
      <label
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          className="ProfilePdfDialog-section"
          defaultChecked={true}
          name="notes"
          style={
            Object {
              "marginRight": 5,
            }
          }
          type="checkbox"
          value="notes"
        />
        Notes
      </label>
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "height": "1.8em",
          "verticalAlign": "middle",
          "width": "12em",
        }
      }
    >
      <label
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          className="ProfilePdfDialog-section"
          defaultChecked={true}
          name="services"
          style={
            Object {
              "marginRight": 5,
            }
          }
          type="checkbox"
          value="services"
        />
        Services
      </label>
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "height": "1.8em",
          "verticalAlign": "middle",
          "width": "12em",
        }
      }
    >
      <label
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          className="ProfilePdfDialog-section"
          defaultChecked={true}
          name="attendance"
          style={
            Object {
              "marginRight": 5,
            }
          }
          type="checkbox"
          value="attendance"
        />
        Attendance
      </label>
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "height": "1.8em",
          "verticalAlign": "middle",
          "width": "12em",
        }
      }
    >
      <label
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          className="ProfilePdfDialog-section"
          defaultChecked={true}
          name="discipline_incidents"
          style={
            Object {
              "marginRight": 5,
            }
          }
          type="checkbox"
          value="discipline_incidents"
        />
        Discipline Incidents
      </label>
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "height": "1.8em",
          "verticalAlign": "middle",
          "width": "12em",
        }
      }
    >
      <label
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <input
          className="ProfilePdfDialog-section"
          defaultChecked={true}
          name="assessments"
          style={
            Object {
              "marginRight": 5,
            }
          }
          type="checkbox"
          value="assessments"
        />
        Assessments
      </label>
    </div>
  </div>
  <span
    style={
      Object {
        "fontWeight": "bold",
        "marginBottom": 10,
        "textAlign": "left",
      }
    }
  >
    Select dates for the report:
  </span>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginLeft": 10,
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <label
        style={
          Object {
            "display": "block",
            "marginBottom": 0,
            "width": "4em",
          }
        }
      >
        From:
      </label>
      <div
        className="Datepicker"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <input
          className="datepicker"
          onChange={[Function]}
          style={
            Object {
              "fontSize": 14,
              "padding": 5,
              "width": "8em",
            }
          }
          value="08/15/2016"
        />
      </div>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <label
        style={
          Object {
            "display": "block",
            "marginBottom": 0,
            "width": "4em",
          }
        }
      >
        Until:
      </label>
      <div
        className="Datepicker"
        style={
          Object {
            "display": "block",
          }
        }
      >
        <input
          className="datepicker"
          onChange={[Function]}
          style={
            Object {
              "fontSize": 14,
              "padding": 5,
              "width": "8em",
            }
          }
          value="03/13/2018"
        />
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-end",
      }
    }
  >
    <button
      className="btn btn-warning"
      onClick={[Function]}
      style={
        Object {
          "fontSize": 12,
          "margin": 0,
        }
      }
    >
      Generate Student Report
    </button>
  </div>
</div>
`;
